File: home_assignments/java/TicketSeller/TicketSellerPersistence/src/main/java/ticketseller/persistence/implementations/UserDBRepository.java
package ticketseller.persistence.implementations;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import ticketseller.model.entities.User;
import ticketseller.persistence.JdbcUtils;
import ticketseller.persistence.interfaces.IUserRepository;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Properties;

public class UserDBRepository implements IUserRepository {

    private static final Logger logger = LogManager.getLogger();
    private final JdbcUtils dbUtils;

    public UserDBRepository(Properties props) {
        logger.info("Initializing CarsDBRepository with properties: {} ", props);
        dbUtils = new JdbcUtils(props);
    }

    @Override
    public void add(User user) {
        logger.traceEntry("Saving user {}", user);
        Connection connection = dbUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("insert into Users(username,password) values(?,?)")) {
            preparedStatement.setString(1, user.getUsername());
            preparedStatement.setString(2, user.getPassword());
            int result = preparedStatement.executeUpdate();
            logger.trace("Saved {} instances", result);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB:" + e);
        }
        logger.traceExit();
    }

    @Override
    public void delete(User user) {
        logger.traceEntry("Deleting user {}", user);
        Connection connection = dbUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("delete from Users where username=(?)")) {
            preparedStatement.setString(1, user.getUsername());
            int result = preparedStatement.executeUpdate();
            logger.trace("Deleted {} instances", result);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB:" + e);
        }
        logger.traceExit();
    }

    @Override
    public void update(User user, String username) {
        logger.traceEntry("Updating user with username: {} with info: {}", username, user);
        Connection connection = dbUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("update Users set username=(?),password=(?) where username=(?)")) {
            preparedStatement.setString(1, user.getUsername());
            preparedStatement.setString(2, user.getPassword());
            preparedStatement.setString(3, username);
            int result = preparedStatement.executeUpdate();
            logger.trace("Updated {} instances", result);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB:" + e);
        }
        logger.traceExit();
    }

    /**
     * extracts a user from a result set
     *
     * @param resultSet said set
     * @return said user
     * @throws SQLException - if any error occurs
     */
    private User extractUserFromSet(ResultSet resultSet) throws SQLException {
        logger.traceEntry("Extracting user from set:{}", resultSet);
        String username = resultSet.getString("username");
        String password = resultSet.getString("password");
        User user = new User(username, password);
        logger.traceExit(user);
        return user;
    }

    @Override
    public User findByID(String username) {
        logger.traceEntry("Searching user with username: {}", username);
        User user = null;
        Connection connection = dbUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("select * from Users where username=(?)")) {
            preparedStatement.setString(1, username);
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                if (resultSet.next()) {
                    user = extractUserFromSet(resultSet);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB:" + e);
        }
        logger.traceExit(user);
        return user;
    }

    @Override
    public Iterable<User> findAll() {
        logger.traceEntry();
        Iterable<User> users = getAll();
        logger.traceExit(users);
        return users;
    }

    @Override
    public Collection<User> getAll() {
        logger.traceEntry();
        List<User> users = new ArrayList<>();
        Connection connection = dbUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("select * from Users")) {
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    User user = extractUserFromSet(resultSet);
                    users.add(user);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB:" + e);
        }
        logger.traceExit(users);
        return users;
    }
}

File: home_assignments/csharp/TicketSeller/persistence/implementations/UserDbRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using log4net;
using model.entities;
using persistence.interfaces;
using persistence.utils;

namespace persistence.implementations
{
    public class UserDbRepository : IUserRepository
    {
        private static readonly ILog Logger = LogManager.GetLogger("UserDbRepository");

        private readonly IDictionary<String, string> _props;

        public UserDbRepository(IDictionary<String, string> props)
        {
            Logger.Info("UserDbRepository");
            this._props = props;
        }


        public void Add(User user)
        {
            Logger.InfoFormat("Saving user {0}", user);

            IDbConnection con = DbUtils.GetConnection(_props);

            try
            {
                using (IDbCommand comm = con.CreateCommand())
                {
                    comm.CommandText =
                        "insert into Users(username,password) values(@username,@password)";
                    IDbDataParameter paramUsername = comm.CreateParameter();
                    paramUsername.ParameterName = "@username";
                    paramUsername.Value = user.Username;
                    comm.Parameters.Add(paramUsername);

                    IDbDataParameter paramPassword = comm.CreateParameter();
                    paramPassword.ParameterName = "@password";
                    paramPassword.Value = user.Password;
                    comm.Parameters.Add(paramPassword);

                    int result = comm.ExecuteNonQuery();
                    Logger.InfoFormat("Added {0} users", result);
                }
            }
            catch (Exception e)
            {
                Logger.Error(e);
            }

            Logger.InfoFormat("Exiting Add");
        }

        public void Delete(User user)
        {
            Logger.InfoFormat("Deleting user {0}", user);

            IDbConnection con = DbUtils.GetConnection(_props);

            try
            {
                using (IDbCommand comm = con.CreateCommand())
                {
                    comm.CommandText =
                        "delete from Users where username=@username";

                    IDbDataParameter paramUsername = comm.CreateParameter();
                    paramUsername.ParameterName = "@username";
                    paramUsername.Value = user.Username;
                    comm.Parameters.Add(paramUsername);

                    int result = comm.ExecuteNonQuery();
                    Logger.InfoFormat("Deleted {0} users", result);
                }
            }
            catch (Exception e)
            {
                Logger.Error(e);
            }

            Logger.InfoFormat("Exiting Delete");
        }

        public void Update(User user, string username)
        {
            Logger.InfoFormat("Updating user with username: {0} with info: {1}", username, user);

            IDbConnection con = DbUtils.GetConnection(_props);

            try
            {
                using (IDbCommand comm = con.CreateCommand())
                {
                    comm.CommandText =
                        "update Users set username=@username,password=@password where username=@old_username";

                    IDbDataParameter paramUsername = comm.CreateParameter();
                    paramUsername.ParameterName = "@username";
                    paramUsername.Value = user.Username;
                    comm.Parameters.Add(paramUsername);

                    IDbDataParameter paramPassword = comm.CreateParameter();
                    paramPassword.ParameterName = "@password";
                    paramPassword.Value = user.Password;
                    comm.Parameters.Add(paramPassword);

                    IDbDataParameter paramOldUsername = comm.CreateParameter();
                    paramOldUsername.ParameterName = "@old_username";
                    paramOldUsername.Value = username;
                    comm.Parameters.Add(paramOldUsername);

                    int result = comm.ExecuteNonQuery();
                    Logger.InfoFormat("Updated {0} users", result);
                }
            }
            catch (Exception e)
            {
                Logger.Error(e);
            }

            Logger.InfoFormat("Exiting Update");
        }
        
        private User ExtractUser(IDataReader dataReader)
        {
            Logger.InfoFormat("Extracting user");
            string username = dataReader.GetString(0);
            string password = dataReader.GetString(1);
            User toReturn = new User(username, password);
            Logger.InfoFormat("Extracted user:{0}", toReturn);
            return toReturn;
        }

        public User FindById(string username)
        {
            Logger.InfoFormat("Searching user with id: {0}", username);
            User toReturn = null;
            IDbConnection con = DbUtils.GetConnection(_props);

            try
            {
                using (IDbCommand comm = con.CreateCommand())
                {
                    comm.CommandText =
                        "select username,password from Users where username=@username";
                    IDbDataParameter paramUsername = comm.CreateParameter();
                    paramUsername.ParameterName = "@username";
                    paramUsername.Value = username;
                    comm.Parameters.Add(paramUsername);

                    using (IDataReader dataR = comm.ExecuteReader())
                    {
                        if (dataR.Read())
                        {
                            toReturn = ExtractUser(dataR);
                        }
                    }
                }
            }
            catch (Exception e)
            {
                Logger.Error(e);
            }

            Logger.InfoFormat("Exiting findOne with value {0}", toReturn);
            return toReturn;
        }

        public IEnumerable<User> FindAll()
        {
            Logger.InfoFormat("Finding all festivals");
            IEnumerable<User> toReturn = GetAll();
            Logger.InfoFormat("Exiting FindAll with value {0}", toReturn);
            return toReturn;
        }

        public ICollection<User> GetAll()
        {
            Logger.InfoFormat("Getting users");
            ICollection<User> toReturn = new List<User>();
            IDbConnection con = DbUtils.GetConnection(_props);

            try
            {
                using (IDbCommand comm = con.CreateCommand())
                {
                    comm.CommandText =
                        "select username,password from Users";
                    using (IDataReader dataR = comm.ExecuteReader())
                    {
                        while (dataR.Read())
                        {
                            User user = ExtractUser(dataR);
                            toReturn.Add(user);
                        }
                    }
                }
            }
            catch (Exception e)
            {
                Logger.Error(e);
            }

            Logger.InfoFormat("Exiting GetAll with value {0}", toReturn);
            return toReturn;
        }
    }
}